Additional functions

itfranck opened the pull request:
Connect / Disconnect (manage connecting to lpass)
By exposing the Connect function, the user do not have to know anything about the cli syntax at all...

Register / Unregister
Register-LastPassVault is more complete than Register-SecretVault since we can have specific parameters for command / path without having to go check the doc.

eg:

```
Register-LassPassVault -VaultName MyVault -Command wsl
```
instead of

```
  $Params = @{
        ModuleName = 'SecretManagement.LastPass'
        Name = 'MyVault'
        VaultParameters = @{lpassCommand = wsl}
    }

Register-SecretVault @Params
```
This open the door to adding more CLI element into the module later on.

itfranck committed:
init

itfranck committed:
doc. + small improvements.

itfranck committed:
Header size...

itfranck committed:
No prompt to logout.

itfranck committed:
removed auto-formatting

itfranck commented:
@TylerLeonhardt 
Are you ok with me going ahead and changing `lpassCommand` to `WSL` while at it ?
I don't know about you but I  feel that with the previous PR and this one, we'd be due for a new release soon. 

On my end at last, I feel like what we've done (unless this PR change that) feel very sturdy at this time. 
I am not looking to add anything in  the immediate short term after that. 

So it would make sense, if we are to change lpassCommand to WSL to do it before the next release.

itfranck commented:
On the longer term , things I could see happening: 

- Add PlatyPS (like what I did [here](https://github.com/adamdriscoll/selenium-powershell/tree/master/Help) so the additional functions are documented not in the Readme directly and also that we have the corresponding embedded help. 

- Add more functions to support other CLI operations

- Add alternate connection mode not depending on the CLI. 
Let's face it, the CLI is not really updated and do not change anymore. 

This could potentially be nice to bypass the CLI all together (or not) 

Here's an interesting article about it. Seems detailed too. 

Check this: https://adventures.michaelfbryan.com/posts/lastpass/
Very limited but he based his module on it: https://github.com/ScottEvtuch/PSLastPass

What do you think about that last one ?

itfranck commented:
Back to this PR, I just confirmed the following:

The following option set allow to connect permanently to the vault (without ever asking for password again).
This is perfect for automated testing. 

I'll add ithat to Connect-LastPass right away.
```
wsl lpass login --trust --plaintext-key itfranck@git.com --force
```

TylerLeonhardt commented:
All of this sounds great! I think we should move to wsl in this PR and once this goes in I can do a release.

Now that we have extra cmdlets, yes we should hook up PlatyPS. We'll get there.

I'm weary to remove the dependency on the CLI since it's the officially supported package... But I do like the idea of not needing it so that installation can be easier.

itfranck committed:
Added StayConnected switch parameter.

Since this is a sensitive operation (It saves the decryption key on your hard drive, Force switch also need to be specified).

I used the Convertto-SecuretString -AsplainText -Force approach.

itfranck committed:
Pulled out all error messages.

itfranck committed:
space

itfranck committed:
Fix param name.

itfranck committed:
Update SecretManagement.LastPass.psm1

itfranck committed:
Switched lpassCommand to wsl

itfranck committed:
Made the test work on Windows 5.1

ConvertTo-SecureString -Force is not mandatory anymore for newer version, but is for 5.1

ConvertFrom-SecureString do not have -AsPlainText switch so I create a check and overrided the function. For this reason, I switched Reload.ps1 invoke from a call (child scope) to dot-sourcing (current scope)

itfranck committed:
Sync-LastPassVault + doc fixes

itfranck committed:
Replaced VaultName param. by Vault

Some SecretManagement use Name (Register / Unregister / Test) but others use Vault.

VaultName was a bit too much variety to my taste. :sweat_smile:

itfranck committed:
Sync-LastPassVault Command fix --now is not a thing

Also, sync is lower case.

itfranck committed:
Forgotten VaultName to Vault items...

itfranck committed:
Added sync after Set-Secret in tests

I had a lot of semi-permanent fluke (Windows 5.1).
Adding an extra sync, even though there's one in the Set-Secret code, got rid of the fluke at a 100% rate.

itfranck committed:
Missing considerations for outputType in Register-LastPassVault.

TylerLeonhardt commented:
Let me know when you're ready for a review

itfranck commented:
> I'm weary to remove the dependency on the CLI since it's the officially supported package... But I do like the idea of not needing it so that installation can be easier.

For this bit,  I was thinking of enabling the noCLI experience through an optional vault parameter.
The philosophy being: 
If you have the CLI, then use the CLI.
If you don't want to install it (complicated (at least for Windows user) or just want the "no CLI experience", then you'd have the option. Should that break you can always set your vault to use the CLI again at any point. 

Then, to managed it, you could have the first line of `Invoke-lpass` that check if the noCLI flag is set and call an alternate Invoke-lpassNoCli method if it is.

itfranck commented:
@TylerLeonhardt 
I was letting the PR sit just for a tiny bit to make sure I was not thinking of something or wanted to come back on some details.
We're good now.
Please proceed with the review at your convenience. 😅

itfranck commented on README.md:
```diff
@@ -72,32 +72,40 @@ commands!
 The module also have the following vault parameter, that can be provided at registration.
 
-#### lpassCommand
+#### [switch] wsl
```
Switched all lpassCommand to wsl

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psd1:
```diff
@@ -1,5 +1,5 @@
     ModuleVersion = '1.0'
     RootModule = 'SecretManagement.LastPass.Extension.psm1'
-    FunctionsToExport = @('Set-Secret','Get-Secret','Remove-Secret','Get-SecretInfo','Test-SecretVault')
+    FunctionsToExport = @('Set-Secret','Get-Secret','Remove-Secret','Get-SecretInfo','Test-SecretVault','Invoke-lpass')
```
Invoke-lpass get exposed to the main module but is not exposed to the user.

TylerLeonhardt replied:
I like this for now. If folks want us to export it higher (for like adhoc commands that will be wsl/xplat agnostic) we can revisit.

itfranck commented on SecretManagement.LastPass.psd1:
```diff
@@ -7,6 +7,8 @@
 
 @{
+    # Script module or binary module file associated with this manifest.
+    RootModule        = 'SecretManagement.LastPass.psm1'
```
The main module expose the additional functions

itfranck commented on test/Extension.Tests.ps1:
```diff
@@ -1,6 +1,6 @@
 Describe 'SecretManagement.LastPass tests' {
     BeforeAll {
-        & $PSScriptRoot/reload.ps1
+        . $PSScriptRoot/reload.ps1
```
Since I defined a `ConvertFrom-SecureString` function in Reload.ps1, it need to be dot-sourced so everything is in the current scope.

itfranck commented on test/Extension.Tests.ps1:
```diff
@@ -15,6 +15,7 @@ Describe 'SecretManagement.LastPass tests' {
     It 'Can store a string secret which is treated like a securestring' {
         $secretText = 'This is my string secret'
         Set-Secret -Name $secretName -Vault $VaultName -Secret $secretText
+        Sync-LastPassVault -Vault $VaultName
```
I also had fluke in my tests in a more consistend manner. 
Adding a sync fixed the problem 100% of the time.

We do a sync in Set-Secret too so I wonder if that's not in the way we pick up the secret through Get-SecretInfo and / or there's a processing delay and the second sync simply give some more time to do the deed.

TylerLeonhardt replied:
hmmm I remember seeing this bug where a newly created secret didn't meet the regex we were using... I thought I fixed it but maybe I missed a case.

itfranck replied:
@TylerLeonhardt Yeah, I remember that.

Hmmm... I had that problem initially that if I didn't do an explicit sync in Set-Secret after setting the secret, I would create a duplicate if I create another secret right afterward. 

So I added a `Invoke-lpass 'sync'` at the end of `Set-Secret` and it  resolved the problem.
Now, if I call `Set-Secret` multiple time in a row, it seem to work...

Although.

Using this, I get duplicate created at some point.
This make sense, if we have that kind of behavior in the tests (sync lagging behind) then at some point, we'll get a `Set-Secret` that will not get anything when calling `lpass show` and will create a secret instead of editing the one present, which lead to duplicate.

```

$Vault = @{Vault = 'lpt'; Name = 'ABCDEF' }
1..10 | %{
    Set-Secret @Vault -Secret 'EDCBA' -Verbose 
    Get-Secret @Vault -AsPlainText
    Remove-Secret @Vault 
}


# View duplicate
Get-SecretInfo @vault 
# Cleanup
Get-SecretInfo @vault | % {Remove-Secret -Name $_.Name -Vault $Vault.Vault}

```

As for your concern regarding something missing with `Get-SecretInfo`, I do not think so. 
`Get-Secret` is really simple in comparison and does not filter out secrets based on some pattern.
It does exhibit the same behavior.

itfranck replied:
I don't necessarily see this PR held back because of this though since it is not tied to a change within it. 
I have opened a separate issue : https://github.com/TylerLeonhardt/SecretManagement.LastPass/issues/19#issue-744433733

itfranck commented on test/Extension.Tests.ps1:
```diff
@@ -47,8 +49,9 @@ Describe 'SecretManagement.LastPass tests' {
     It 'Can store a PSCredential secret' {
         $secretText = 'This is my pscredential secret'
-        $secret = [PSCredential]::new('myUser', ($secretText | ConvertTo-SecureString -AsPlainText))
+        $secret = [PSCredential]::new('myUser', ($secretText | ConvertTo-SecureString -AsPlainText -Force))
```
-Force is superfluous for the newer PS version but still mandatory for 5.1. so I added it to be able to test with 5.1 (

itfranck commented on test/reload.ps1:
```diff
@@ -10,6 +17,9 @@ foreach ($module in $modules) {
     }
 }
 
-Register-SecretVault (Join-Path $PSScriptRoot '..' 'SecretManagement.LastPass.psd1') -Name 'LastPass.Tests'
```
This didn't work with 5.1

TylerLeonhardt replied:
Yeah because in 5.1 `Join-Path` only took in 2 paths...

5.1 ... so limiting...

TylerLeonhardt commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -41,7 +41,7 @@ $DefaultNoteTypeMap = @{
     AccountNotFound = 'Error: Could not find specified account(s).'
     # Need to use wildcard since the path of lpass could be different
-    LoggedOut = 'Error: Could not find decryption key. Perhaps you need to login with*'
+    LoggedOut = 'Error: Could not find decryption key. Perhaps you need to login with Connect-LastPass'
```
Why did you make this change? That star is needed to properly surface the right error.

itfranck replied:
Indeed.
I'll revert it.
I missed the `-wildcard` and the fact it was not only the output message but also used in the switch statement. 

```suggestion
    LoggedOut = 'Error: Could not find decryption key. Perhaps you need to login with*'
```


That being said, I just found another thing. 
Since we do not redirect error stream anymore on the Get so we get the prompt, we won't get that message displayed ever again. 

(Because $result will always be $null so I am keeping this conversation open and I will review if we can indicate to the user it need to login instead of displaying: The secret <name> was not found **Figure-2)** 
![image](https://user-images.githubusercontent.com/1980296/99342250-78463700-2859-11eb-90f6-b54cd891a9bd.png)

**Figure-2**
![image](https://user-images.githubusercontent.com/1980296/99342416-cc511b80-2859-11eb-8126-6f6d56e0d0b8.png)

TylerLeonhardt replied:
@itfranck I guess I kept it in in the event that the user has _never_ logged in. Prompting for the password requires you to have logged in in the past AFAICT because otherwise it would need to prompt for username as well ... but maybe I have this wrong.

itfranck replied:
I've created an issue and will deal with it specifically in its own PR so I can close this one. 

I think I have a good idea how to fix this. 
Will elaborate soon.

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    $Arguments = [System.Collections.Generic.List[String]]@('login')
+    if ($trust) { $Arguments.Add('--trust') }
+    if ($StayConnected) { 
+        if (! $Force) { Throw [System.Management.Automation.PSArgumentException] $ErrorMessages.StayConnectedForceSwitchMissing }
```
This should be implemented using ShouldProcess mechanics:

https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-shouldprocess?view=powershell-7.1#implementing--force

itfranck replied:
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/18/commits/2fc9336324651b26c959788b9c8cdd0b4f497d09

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    Unregister_NotLpass_1           = "The specified vault is not a $ModuleName vault (VaultType: {0}"
+}
+
+function Connect-LastPass {
```
Can you add basic comment-based help? In vscode (with PS ext), if you type `##` above or below the `function Connect-LastPass` line, it should scaffold it for you.

itfranck replied:
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/18/commits/563afb321f386e4e23f2f563f3c09d0d1bfdb4e7

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+        [String]$Path
+    )
+
+    $Params = @{
+        ModuleName      = 'SecretManagement.LastPass'
+        Name            = if ('' -ne $Vault) {$Vault} else {$ModuleName}
+        VaultParameters = @{}
+    }
+    if ($wsl -eq $true) { $Params.VaultParameters.Add('wsl', $true) }
+    if ($Path -ne '') { $Params.VaultParameters.Add('lpassPath', $Path) }
+    if ($Detailed -eq $true) { $Params.VaultParameters.Add('outputType','Detailed') }
+    if ($VerbosePreference -eq 'Continue') {$Params.add('Verbose',$true)}
```
```suggestion
    $Params = @{
        ModuleName      = 'SecretManagement.LastPass'
        Name            = if ('' -ne $Vault) {$Vault} else {$ModuleName}
        Verbose         = $VerbosePreference -eq 'Continue'
        VaultParameters = @{
            wsl         = $wsl.IsPresent
            outputType  = $Detailed.IsPresent
        }
    }

    if ($Path -ne '') { $Params.VaultParameters.Add('lpassPath', $Path) }
```

a tad cleaner IMO?

itfranck replied:
I couldn't take the suggestion quite as it was presented because of outputType. 
It is not a boolean. 

```
$Params = @{
        ModuleName      = 'SecretManagement.LastPass'
        Name            = if ('' -ne $Vault) {$Vault} else {$ModuleName}
        Verbose         = $VerbosePreference -eq 'Continue'
        VaultParameters = @{
            wsl         = $Wsl.IsPresent
            outputType  = if ($Detailed) { 'Detailed' } else { 'Default' }
        }
    }
```

https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/18/commits/1d0107c5bc256437e32f77d56a49f9a0aeec9ae4

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    if ($VerbosePreference -eq 'Continue') {$Params.Add('Verbose',$true)}
+    
+    $Vault = Get-SecretVault -Name $params.Name -ErrorAction Stop
+    if ($Vault.ModuleName -ne $ModuleName) { Throw $ErrorMessages.Unregister_NotLpass_1 -f $Vault.ModuleName}
```
```suggestion
    if ($Vault.ModuleName -ne $ModuleName) { Throw $ErrorMessages.Unregister_NotLpass_1 -f $Vault.ModuleName }
```

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    param (
+        [String]$Vault
+    )
+    $Params = @{Name = if ('' -ne $Vault) { $Vault } else { $ModuleName } }
+    if ($VerbosePreference -eq 'Continue') {$Params.Add('Verbose',$true)}
```
```suggestion
    $Params = @{
        Name = if ('' -ne $Vault) { $Vault } else { $ModuleName }
        Verbose = $VerbosePreference -eq 'Continue'
    }
```

a bit cleaner

TylerLeonhardt replied:
don't forget this one?

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    $VaultParams = Get-VaultParams -Vault $Vault
+    Invoke-lpass -Arguments 'sync' -VaultParams $VaultParams
+}
+Function Get-VaultParams {
```
```suggestion

function Get-VaultParams {
```

whitespace and consistant style

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+Function Get-VaultParams {
+    Param($Vault)
+    if ([String]::IsNullOrEmpty($Vault)) { 
+        $AllVaults = Get-SecretVault | Where-Object ModuleName -eq $ModuleName
```
wouldn't it be best to use `-Name` on `Get-SecretVault` and then handle when there isn't one called `$ModuleName`?

itfranck replied:
Done.
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/18/commits/29e86919b71bd42ef6501ed722ba3764d3ae3457

TylerLeonhardt replied:
do we need the switch anymore? Can you register the same vault name twice?

itfranck replied:
@TylerLeonhardt 
You cannot register the same vault twice. 
The behavior is this.

- If user do not provide a vault name (`[String]::IsNullOrEmpty($Vault)`) 
- We get the vault called SecretManagement.LastPass and return its parameter
BUT

If there's no vault with such name, the switch statement kick in and return the vault parameter of any `SecretManagement.LastPass` vault under the condition that there is only 1. 

If there's a different number, an error is thrown to ask the user to provide a vault name (multiple) or to indicate there's no lastpass vault at all. 

Get-VaultParams is actually not used at all by Register / Unregister

TylerLeonhardt replied:
ahhh I see. That makes sense! I didn't see the _ModuleName_ in the where clause and thought you had specified Name. Maybe worth a comment.

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -0,0 +1,114 @@
+    [CmdletBinding()]
+    param (
+        [String]$Vault,
+        [switch]$wsl,
```
I think this should be `Wsl` to be consistant with the others... but the vault parameter can still be `wsl`

TylerLeonhardt reviewed (commented):
Looking so good! This is going to be the best vault out there 😅

itfranck committed:
Revert logged out message

itfranck committed:
Update SecretManagement.LastPass.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Update SecretManagement.LastPass.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
wsl to Wsl for function parameter

itfranck committed:
Update SecretManagement.LastPass.psm1

https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/18#discussion_r524845246

itfranck committed:
A tad cleaner IHO

itfranck committed:
Connect-LastPass ShouldProcess

itfranck committed:
Basic comment-based help.

itfranck commented:
Ready again. as per comments, I created a separate issue for the comment you made about the extra-sync.
It looks like there's an issue but this is also reproducing by calling Get-Secret immetidately after Set-Secret and thus, I don't think you missed something in Get-SecretInfo.

TylerLeonhardt commented:
commented on a couple other things. They might be kinda buried but I unresolved the ones I replied to.

itfranck committed:
Update SecretManagement.LastPass.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

TylerLeonhardt reviewed (approved):
LGTM :) you should have rights to the repo so merge it in when you're ready (if you decide to add that comment)

itfranck committed:
Comment.

itfranck commented:
@TylerLeonhardt I'm upset.
I did a squash & merge but it failed.
I clicked Try again, refreshed the page and it said "Merged"
It indeed merged. 

It's the first time I get something like that.

Anything you want me to do to undo the many commits into main or is it good ? 
(From 13 to 42 😞 )

TylerLeonhardt commented:
Eh it's ok
